pallets/utxo/src/staking.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2021 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): C. Yap
+
+use crate::{Config, StakingCount, Error, Event, Pallet, TransactionOutput, Destination, LockedUtxos, Value, UtxoStore, MLTCoinsAvailable};
+use frame_support::{dispatch::{DispatchResultWithPostInfo, Vec}, ensure, traits::Get};
+use sp_core::H256;
+use sp_std::vec;
+use sp_runtime::transaction_validity::{ValidTransaction,TransactionLongevity};
+use sp_runtime::traits::{BlakeTwo256, Hash};
+
+/// A helper trait to handle staking NOT found in pallet-utxo.
+pub trait StakingHelper<AccountId>{
+
+    /// to convert a public key into an AccountId
+    fn get_account_id(pubkey: &H256) -> AccountId;
+
+    /// start the staking.
+    /// # Arguments
+    /// * `stash_account` - A placeholder of the "supposed" validator. This is only to "satisfy"
+    /// the `pallet-staking`'s needs to be able to stake.
+    /// * `controller_account` - The ACTUAL validator. But this is NOT SO, in the `pallet-staking`.
+    /// In `pallet-staking`, its job is like an "accountant" to the stash account.
+    /// * `session_key` - to get up-to-date with validators, eras, sessions. see `pallet-session`.
+    /// * `value` - the amount to stake/bond/stash
+    fn stake(stash_account:&AccountId, controller_account:&AccountId, session_key:&mut Vec<u8>, value:Value) -> DispatchResultWithPostInfo;
+
+    /// stake more funds for the validator
+    fn stake_extra(controller_account:&AccountId, value:Value) -> DispatchResultWithPostInfo;
+
+    /// quitting the role of a validator.
+    fn pause(controller_account:&AccountId) -> DispatchResultWithPostInfo;
+
+    /// transfer balance from the locked state to the actual free balance.
+    fn withdraw(controller_account: &AccountId) -> DispatchResultWithPostInfo;
+
```

Comment:
I kind of imagined the reward would go to the stash to be reused for staking. If one puts the stash as payee, will that reward be restaked?